sdks/python/apache_beam/runners/interactive/caching/streambasedcache.py

```diff
@@ -0,0 +1,491 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from __future__ import absolute_import
+
+import contextlib
+import functools
+import time
+import uuid
+from datetime import datetime
+
+import pytz
+from future.moves import queue
+
+import apache_beam as beam
+import apache_beam.io.gcp.pubsub as beam_pubsub
+from apache_beam import coders
+from apache_beam.io.gcp.pubsub import PubsubMessage
+from apache_beam.io.gcp.pubsub import ReadFromPubSub
+from apache_beam.io.gcp.pubsub import WriteToPubSub
+from apache_beam.runners.direct.direct_runner import _DirectWriteToPubSubFn
+from apache_beam.runners.interactive.caching import pcollectioncache
+from apache_beam.testing import datatype_inference
+from apache_beam.transforms import PTransform
+from apache_beam.transforms.window import TimestampedValue
+from apache_beam.utils.timestamp import Timestamp
+
+try:
+  from weakref import finalize
+except ImportError:
+  from backports.weakref import finalize
+
+try:
+  from google import api_core
+  from google.cloud import pubsub
+  from google.api_core import exceptions as gexc
+except ImportError:
+  api_core = None
+  pubsub = None
+  gexc = None
+
+__all__ = [
+    "StreamBasedCache",
+    "PubSubBasedCache",
+]
+
+
+class StreamBasedCache(pcollectioncache.PCollectionCache):
+  pass
+
+
+class PubSubBasedCache(StreamBasedCache):
```

Comment:
Could you an explanation on how this cache works?
- How cached data is encoded.
- Where cached data is written to.
- How snapshots are used. What happens when snapshots are supported, and what happens when they are not.
- What with_attributes does.
- The special "ts" attribute.

Reply:
I covered each of those points in the docstring for `PubSubBasedCache`. Thanks for the detailed list!